### PR TITLE
[CUDA] CUDA Quantized Training (fixes #5606)

### LINKED ISSUE #5606: Add quantized training
## Summary

Add quantized training algorithm, which accelerates training on both CPU and GPU. See our latest NeurIPS 2022 paper (https://neurips.cc/virtual/2022/poster/53370).

### PR DESCRIPTION
**Pull request by shiyu1994**
Fixes #5606.

Adds quantized training for CUDA version.

### CODE REVIEW

**Comment by guolinke** [comment]
@shiyu1994 is this ready for review?

**Comment by shiyu1994** [comment]
@guolinke This is ready. Please check.

File: src/treelearner/cuda/cuda_best_split_finder.cpp [hunk 1]
```diff
@@ -40,6 +40,9 @@ CUDABestSplitFinder::CUDABestSplitFinder(
   select_features_by_node_(select_features_by_node),
   cuda_hist_(cuda_hist) {
   InitFeatureMetaInfo(train_data);
+  if (has_categorical_feature_ && config->use_quantized_grad) {
```
Comment (shiyu1994) [inline src/treelearner/cuda/cuda_best_split_finder.cpp:43]: Link #6119

**Comment by shiyu1994** [comment]
@jameslamb I've enlarged the size limitation for distributed package to 100M. Because we add a few more templates in the PR which add to the size of compiled file. Do you think it is OK?

**Comment by jameslamb** [comment]
> I've enlarged the size limitation for distributed package to 100M. Because we add a few more templates in the PR which add to the size of compiled file. Do you think it is OK?

Thanks for the `@`.

For now, since we're not distributing these CUDA wheels on PyPI, I think it's ok. Let's not let it block this PR.

But if we pursue shipping a fat wheel in the future with CUDA support precompiled (like we talked about in Slack), 100MB will be a problem.

There are limits on PyPI for both individual file size and cumulative project size. I don't know the exact numbers but shipping 100MB wheels would put us in the range of hitting them, I think.

See these discussions:

* https://github.com/pypi/support/issues/2341
* https://discuss.python.org/t/what-to-do-about-gpus-and-the-built-distributions-that-support-them/7125

There are also other concerns with such large wheels, e.g. for people using function-as-a-service things like AWS Lambda. See for example:

* https://github.com/pandas-dev/pandas/issues/30741

I'll open a new issue in the next few days to discuss publishing wheels with CUDA support.

**Comment by jameslamb** [comment]
I removed the `feature` label from this and left `efficiency`. For release-drafter, I think it can only be one of the labels specified here, not multiple:

https://github.com/microsoft/LightGBM/blob/f175cebd2d971ca6140455035e526d67f04237eb/.github/release-drafter.yml#L3-L15

**Review by guolinke — APPROVED** [review]
Thank you!

**Pull request merged**